meteor_packages/mats-common/public/python/mode_stats.py

```diff
@@ -0,0 +1,313 @@
+import numpy as np
+
+error = ""
+
+
+# helper function for MODE calculations
+def get_interest2d(sub_interest, sub_mode_header_id, sub_pair_fid, sub_pair_oid):
+    # set up 2-dimensional interest arrays for calculating MODE stats
+    # need a separate 2d array for each mode_header_id
+    unique_mode_headers = list(set(sub_mode_header_id))
+    mode_header_lookup = {}
+    mode_index_lookup = {}
+    interest_2d_arrays = {}
+    for unique_header in unique_mode_headers:
+        mode_header_lookup[str(unique_header)] = {
+            "lookup_f_index": 0,
+            "lookup_o_index": 0
+        }
+    for i in range(0, len(sub_interest)):
+        # create indices for different mode_header_ids
+        this_mode_header_id = str(sub_mode_header_id[i])
+        if this_mode_header_id not in mode_index_lookup.keys():
+            mode_index_lookup[this_mode_header_id] = {}
+        this_fid = sub_pair_fid[i]
+        this_oid = sub_pair_oid[i]
+        if this_fid not in mode_index_lookup[this_mode_header_id].keys():
+            mode_index_lookup[this_mode_header_id][this_fid] = \
+                mode_header_lookup[this_mode_header_id]["lookup_f_index"]
+            mode_header_lookup[this_mode_header_id]["lookup_f_index"] = \
+                mode_header_lookup[this_mode_header_id]["lookup_f_index"] + 1
+        if this_oid not in mode_index_lookup[this_mode_header_id].keys():
+            mode_index_lookup[this_mode_header_id][this_oid] = \
+                mode_header_lookup[this_mode_header_id]["lookup_o_index"]
+            mode_header_lookup[this_mode_header_id]["lookup_o_index"] = \
+                mode_header_lookup[this_mode_header_id]["lookup_o_index"] + 1
+    for unique_header in unique_mode_headers:
+        this_mode_header_id = str(unique_header)
+        interest_2d_arrays[this_mode_header_id] = \
+            np.zeros((mode_header_lookup[this_mode_header_id]["lookup_f_index"],
+                      mode_header_lookup[this_mode_header_id]["lookup_o_index"]), dtype=np.float)
+    for i in range(0, len(sub_interest)):
+        this_mode_header_id = str(sub_mode_header_id[i])
+        this_fid = sub_pair_fid[i]
+        this_oid = sub_pair_oid[i]
+        interest_2d_arrays[this_mode_header_id][mode_index_lookup[this_mode_header_id][this_fid],
+                                                mode_index_lookup[this_mode_header_id][this_oid]] = sub_interest[i]
+    return interest_2d_arrays, mode_header_lookup
+
+
+# helper function for MODE calculations
+def gc_dist(lon1, lat1, lon2, lat2):
+    r_e = 6.371e6  # [m]
```

Comment:
That is a helper function from Jeff Duda's code to calculate geometric distance on a sphere between two points. We didn't end up using it in the mean centroid distance algorithm he ultimately chose, but I've left it in in case he changes his mind. If he hasn't in two more METexpress versions then I will remove it.

Reply:
Makes sense, thanks! If we do decide to use it, it'd be good to document what `r_e` represents.